clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -843,7 +845,11 @@ private void updatePatternSubscription(Cluster cluster) {
     @Override
     public void unsubscribe() {
```

Comment:
There is a subtile behaviour changes here.

1) In the legacy implementation, `this.coordinator.onLeavePrepare()` is called here and it triggers the callback before returning from `unsubscribe`.

2) `subscriptions.unsubscribe()` is actually called before `unsubscribe` returns as well.

Reply:
I updated it to align with the current behaviour (callbacks, best effort to send leave group request without any response handling or retry, and call to `subscriptions.unsubscribe` when everything completes). This has the gap of the callback execution that would require a poll. Given that we don't support callbacks in this PR, it won't block the flow, but definitely to be solved (I added the details of the challenge to solve in the [callbacks Jira](https://issues.apache.org/jira/browse/KAFKA-15276))